### PR TITLE
kubenet checks required cni plugins are present

### PR DESCRIPTION
**Pull request by freehan**
related to #31215

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31931)

### CODE REVIEW

File: pkg/kubelet/network/kubenet/kubenet_linux.go [hunk 1]
```diff
@@ -531,9 +534,43 @@ func (plugin *kubenetNetworkPlugin) Status() error {
 	if plugin.netConfig == nil {
 		return fmt.Errorf("Kubenet does not have netConfig. This is most likely due to lack of PodCIDR")
 	}
+
+	if !plugin.checkCNIPlugin() {
+		return fmt.Errorf("Could not locate kubenet required CNI plugins %v at %s or %s", requiredCNIPlugins, DefaultCNIDir, plugin.vendorDir)
```
Comment (yujuhong) [inline pkg/kubelet/network/kubenet/kubenet_linux.go]: nit: s/Could/could
s/at %s or %s/at %q or %q

File: pkg/kubelet/network/kubenet/kubenet_linux.go [hunk 2]
```diff
@@ -531,9 +534,43 @@ func (plugin *kubenetNetworkPlugin) Status() error {
...
+	}
 	return nil
 }
 
+// checkCNIPlugin returns if all kubenet required cni plugins can be found at /opt/cni/bin or user specifed NetworkPluginDir.
+func (plugin *kubenetNetworkPlugin) checkCNIPlugin() bool {
```
Comment (yujuhong) [inline pkg/kubelet/network/kubenet/kubenet_linux.go]: nit: seems like this function can just return the error in line 539

**Comment by yujuhong** [comment]
@dchen1107 @wojtek-t this PR will ensure kubelet to post NodeNotReady if the required network plugins are missing. (It doesn't deal with wrong permission or corruption).

**Comment by dchen1107** [comment]
I am ok to mark NodeNotReady with proper error message indicating why. Meanwhile, I think we should re-try to download those tarball at the first place. Of course that is out-of-scope of this pr.

**Comment by dchen1107** [comment]
Can I have a test case to reproduce the issue and check the nodestatus? I can have that in a separate pr. 

LGTM

File: pkg/kubelet/network/kubenet/kubenet_linux.go [hunk 3]
```diff
@@ -531,9 +534,43 @@ func (plugin *kubenetNetworkPlugin) Status() error {
...
+	if plugin.checkCNIPluginInDir(DefaultCNIDir) || plugin.checkCNIPluginInDir(plugin.vendorDir) {
+		return true
+	}
+	return false
+}
+
+// checkCNIPluginInDir returns if all required cni plugins are placed in dir
+func (plugin *kubenetNetworkPlugin) checkCNIPluginInDir(dir string) bool {
+	output, err := plugin.execer.Command("ls", dir).CombinedOutput()
```
Comment (luxas) [inline pkg/kubelet/network/kubenet/kubenet_linux.go]: Argh, no more `exec`ing should be allowed :)
Please rewite this with native go tools in the future.

This is a much better and readable approach for example: http://stackoverflow.com/a/14668907

Comment (freehan) [inline pkg/kubelet/network/kubenet/kubenet_linux.go]: I poked around yesterday for a better util for this. Will submit a patch while I am still on it.

**Comment by wojtek-t** [comment]
@freehan @dchen1107 @yujuhong - thanks a lot for this PR!

**Pull request merged**